compiler/luci/pass/src/SubstituteSqueezeToReshapePass.cpp

```diff
@@ -0,0 +1,166 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/SubstituteSqueezeToReshapePass.h"
+
+#include <luci/IR/CircleNodes.h>
+
+namespace
+{
+
+/**
+ * @brief Return TRUE if all dim is known && dim > 0 && at least dim 1 exist
+ */
+bool can_squeeze_shape(luci::CircleNode *node)
+{
+  int dim_1_cnt = 0;
+  for (uint32_t r = 0; r < node->rank(); ++r)
+  {
+    if (not node->dim(r).known())
+      return false;
+    if (node->dim(r).value() < 1)
+      return false;
+    if (node->dim(r).value() == 1)
+      dim_1_cnt++;
+  }
+  return dim_1_cnt > 0;
+}
+
+/**
+ * @brief return shape vector of input excluding squeeze_dims
+ */
+std::vector<uint32_t> new_shape_with_squeeze(const luci::CircleNode *input,
+                                             const std::vector<int32_t> &squeeze_dims)
+{
+  std::vector<uint32_t> new_shape;
+
+  if (squeeze_dims.size() == 0)
+  {
+    // prepare new_shape from input shape excluding dim 1
+    for (uint32_t r = 0; r < input->rank(); ++r)
+    {
+      auto v = input->dim(r).value();
+      if (v != 1)
+        new_shape.push_back(v);
+    }
+  }
+  else
+  {
+    // prepare new_shape from input shape excluding squeeze_dims
+    std::vector<uint32_t> temp_shape;
+    for (uint32_t r = 0; r < input->rank(); ++r)
+      temp_shape.push_back(input->dim(r).value());
+    for (auto dim : squeeze_dims)
+    {
+      assert(temp_shape.at(dim) == 1);
+      temp_shape.at(dim) = 0;
+    }
+    for (auto ts : temp_shape)
+      if (ts > 0)
+        new_shape.push_back(ts);
+  }
+
+  // TODO check if new_shape is empty
+
+  return new_shape;
+}
+
+/**
+ * @brief return CircleConst ptr with values of new_shape
+ */
+luci::CircleConst *create_shape_const(loco::Graph *graph, const std::vector<uint32_t> &new_shape)
+{
+  uint32_t dim_size = static_cast<uint32_t>(new_shape.size());
+
+  auto shape_const = graph->nodes()->create<luci::CircleConst>();
+
+  // const shape/dtype
+  shape_const->dtype(loco::DataType::S32);
+  shape_const->shape({dim_size});
+  shape_const->shape_status(luci::ShapeStatus::VALID);
+
+  // constant values
+  shape_const->size<loco::DataType::S32>(dim_size);
+  for (uint32_t i = 0; i < new_shape.size(); ++i)
+    shape_const->at<loco::DataType::S32>(i) = new_shape.at(i);
+
+  return shape_const;
+}
+
+bool substitute_squeeze_to_reshape(luci::CircleSqueeze *squeeze)
+{
+  assert(squeeze != nullptr);
+
+  auto input = loco::must_cast<luci::CircleNode *>(squeeze->input());
+  // we need input node shape and all dim should be known
+  if (input->shape_status() != luci::ShapeStatus::VALID)
+    return false;
+  if (not can_squeeze_shape(input))
```

Comment:
Following case cannot be optimized by this condition.
```python
inp = tf.constant(1, shape=[2,3,4,5])
sq = tf.squeeze(inp)

# Expected optimized result : tf.reshape(inp, [2,3,4,5])
# Obtained optimized result : Not optimized
```

Reply:
Um? can you explain more about expected result?

Reply:
```python
inp = tf.constant(1, shape=[2,3,4,5])
sq = tf.squeeze(inp)
```

If input of `Squeeze` has [2, 3, 4, 5] shape, any dimension will be squeezed.
So the result of `sq` would be same with `inp`.
It means that the `Squeeze` can be substituted by `Reshape` with new_shape [2, 3, 4, 5].

Reply:
> any dimension will be squeezed.

there is no `1` in input shape. so, I don't think there will be any shape change
If I'm not correct, please let me know.

> Squeeze can be substituted by Reshape with new_shape [2, 3, 4, 5].

OK, now I understand what you mean. For this case, Squeeze will not be doing what Squeeze  supposed to do.
So we can just replace this Reshape and then later it can be removed.
Is your suggestion like this?

Reply:
> there is no 1 in input shape. so, I don't think there will be any shape change
> I'm not correct, please let me know.

You are correct :)

> OK, now I understand what you mean. For this case, Squeeze will not be doing what Squeeze is supposed to do.
> Squeeze is doing nothing, like a dummy. So we can just replace this Reshape and then later it can be removed.
> Is your suggestion like this?

Exactly :D